relays/equivocation/src/equivocation_loop.rs

```diff
@@ -0,0 +1,327 @@
+// Copyright 2019-2023 Parity Technologies (UK) Ltd.
+// This file is part of Parity Bridges Common.
+
+// Parity Bridges Common is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Parity Bridges Common is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Parity Bridges Common.  If not, see <http://www.gnu.org/licenses/>.
+
+use crate::{
+	reporter::EquivocationsReporter, EquivocationDetectionPipeline, HeaderFinalityInfo,
+	SourceClient, TargetClient,
+};
+
+use bp_header_chain::{FinalityProof, FindEquivocations};
+use finality_relay::{FinalityProofsBuf, FinalityProofsStream};
+use futures::{select, FutureExt};
+use num_traits::Saturating;
+use relay_utils::{
+	relay_loop::{reconnect_failed_client, RECONNECT_DELAY},
+	FailedClient, MaybeConnectionError,
+};
+use std::{future::Future, time::Duration};
+
+struct EquivocationReportingContext<P: EquivocationDetectionPipeline> {
```

Comment:
Could do with some descriptive docs for all structs and methods throughout this file